### PR TITLE
Initial basic training template (densenet / cifar10)

### PR DESCRIPTION
**Pull request by ianjjohnson**
# What does this PR do?

This creates an initial template for training KerasCV model architectures.
It currently supports
- Failure recovery
- Storing weights to GCS
- Basic tensorboard output to GCS
- A simple custom learning rate schedule
- Very basic metadata output

It's not ready for real use at this point, but let's talk through the structure and high-level approach, and then we can add features from here.

Some open questions:
- How generic do we want to make training scripts (I'd like to avoid too much code duplication but also recognize that it's important to avoid unnecessary indirection, especially since these should mostly stand alone)
- Should we have a single copy of the model setup / training / weight storage code, and make each training script implement just data augmentation and a training loop? Or should we embrace some code duplication so that training scripts can be more easily customized

@LukeWood @tanzhenyu

### CODE REVIEW

File: examples/training/classification/densenet.py [hunk 1]
```diff
@@ -53,6 +68,7 @@ def main(argv):
     gcs_tensorboard_path = gcs_path_base + "logs/"
 
     train, test = load_cfar10_dataset(BATCH_SIZE)
+    train = train.map(augment)
```
Comment (LukeWood) [inline examples/training/classification/densenet.py:72]: add num_parallel_calls=tf.data.AUTOTUNE to map calls

File: examples/training/classification/densenet.py [hunk 2]
```diff
@@ -53,6 +68,7 @@ def main(argv):
     gcs_tensorboard_path = gcs_path_base + "logs/"
 
     train, test = load_cfar10_dataset(BATCH_SIZE)
```
Comment (LukeWood) [inline examples/training/classification/densenet.py:71]: cfar => cifar

File: examples/training/classification/densenet.py [hunk 3]
```diff
@@ -36,10 +37,24 @@
...
-EPOCHS = 10
+EPOCHS = 250
 BATCH_SIZE = 32
 WEIGHTS_PATH = "weights.hdf5"
 
+AUGMENT_LAYERS = [
+    keras_cv.layers.RandAugment(value_range=(0, 255), magnitude=0.7),
+    keras_cv.layers.CutMix(),
+    keras_cv.layers.MixUp(),
+    keras_cv.layers.RandomFlip(),
```
Comment (LukeWood) [inline examples/training/classification/densenet.py:48]: believe it or not, order matters.

RandomFlip() => first
RandAugment() => second
RandomCutout() => third
CutMix() => fourth
MixUp => fifth

File: examples/training/classification/densenet.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    # In case storing weights in GCS fails, the weights are also stored locally.
+    os.system(
+        "gsutil cp {local_path} {remote_path}".format(
+            local_path=WEIGHTS_PATH, remote_path=gcs_weights_path
+        )
+    )
+
+    metadata = {
+        "experiment_id": _EXPERIMENT_ID.value,
+        "gcs_weights_path": gcs_weights_path,
```
[the remark this answers is not in the document]
Comment (ianjjohnson) [reply examples/training/classification/densenet.py:133]: I've added author, and a dict of evaluation metrics (which is currently just accuracy)

File: examples/training/classification/densenet.py (hunk 3 above)
Comment (ianjjohnson) [reply examples/training/classification/densenet.py]: Good to know -- is there a good reference I can look at about why order is important here?

Comment (ianjjohnson) [reply examples/training/classification/densenet.py]: (Such a reference would probably be good to point to in the training script)

Comment (LukeWood) [reply examples/training/classification/densenet.py:54]: not that I know of actually... 🤔

**Comment by ianjjohnson** [comment]
I am closing this while we think more about how these scripts should be structured.
